### PR TITLE
Upgrade python 3

### LINKED ISSUE #60: Python 3 support
Thanks for this repo. Do you have a rough guess of what it would take to support python 3?

### PR DESCRIPTION
**Pull request by wcedmisten**
WIP branch for upgrading to Python 3. Currently training does not work and will crash the image.

Python 3 Upgrade Summary:

* `u"string"` syntax is obsolete because strings are unicode by default 
* imports must be absolute
* csv library uses unicode and can replace `unicodecsv` import

One problem I'm running into is that training the model in the docker image will crash the docker image and the host computer. Unclear why this is happening right now.

Fixes #60

### CODE REVIEW

**Comment by mtlynch** [comment]
I can't figure out why, but Travis CI was refusing to build your PR. I switched to Circle CI. (#65)

Could you rebase onto master so that CI starts building? It'll help me give you feedback on possible causes for the break.

**Comment by mtlynch** [comment]
It's currently failing on the formatting checker. You can reformat with this command:

```bash
yapf \
  --in-place \
  --recursive \
  --style google \
  ./ \
  --exclude="third_party/*" \
  --exclude="build/*"
```

**Comment by wcedmisten** [comment]
I believe the CI pipeline is now reproducing the error I was getting on my machine.

**Comment by wcedmisten** [comment]
Update, it looks like the csvWriter is writing bytestrings (`b""`) as output which is causing CRF++ to fail. I'll keep looking into this.

**Comment by dgrant** [comment]
@wcedmisten Why not just run `2to3` on all the python code? It seems you did things manually as the changes are spread across a few commits.

**Comment by wcedmisten** [comment]
> @wcedmisten Why not just run `2to3` on all the python code? It seems you did things manually as the changes are spread across a few commits.

I did use this tool to find which areas of code needed updating, but decided to commit the changes categorized by type.

**Comment by mtlynch** [comment]
Just a heads up, I'll accept a PR that migrates from Python2 to Python3, but the PR should exclusively perform that migration and not otherwise change behavior that affects existing clients of this tool or reorganize the code beyond what's necessary for the 2-to-3 migration.

I'm noticing that 69cead3 is adding a major change to the CLI that will break existing clients, and another commit unnecessarily [moves the `tokenize` function](https://github.com/mtlynch/ingredient-phrase-tagger/pull/62/files#diff-c55ddf61c3c836800aa2f1ab5709bc10fc789ca637b77d2363dca33b077cdc80R299) out of the `tokenizer` module and into the poorly-named `utils` module. Those changes are unrelated to the 2 to 3 migration, so they shouldn't be part of this PR.

I just wrote an article about [best practices for submitting PRs](https://mtlynch.io/code-review-love/) that may help clarify my expectations here so that I can get this merged in for you as smoothly as possible.

**Comment by wcedmisten** [comment]
> Just a heads up, I'll accept a PR that migrates from Python2 to Python3, but the PR should exclusively perform that migration and not otherwise change behavior that affects existing clients of this tool or reorganize the code beyond what's necessary for the 2-to-3 migration.
> 
> I'm noticing that [69cead3](https://github.com/mtlynch/ingredient-phrase-tagger/commit/69cead370382705c53a3a2b3c66571ad74ed566e) is adding a major change to the CLI that will break existing clients, and another commit unnecessarily [moves the `tokenize` function](https://github.com/mtlynch/ingredient-phrase-tagger/pull/62/files#diff-c55ddf61c3c836800aa2f1ab5709bc10fc789ca637b77d2363dca33b077cdc80R299) out of the `tokenizer` module and into the poorly-named `utils` module. Those changes are unrelated to the 2 to 3 migration, so they shouldn't be part of this PR.
> 
> I just wrote an article about [best practices for submitting PRs](https://mtlynch.io/code-review-love/) that may help clarify my expectations here so that I can get this merged in for you as smoothly as possible.

Wow! I read this article on /r/programming last week, what a coincidence! Thanks for the feedback, I agree with all your points. This was a good reminder to keep general improvement refactoring in a separate PR. (So you might get some of those, too).

I reverted the change that consolidated tokenizer.py into utils.py. That was originally done to remove the circular imports between those two files, which was related to the 2 to 3 upgrade, but I fixed this instead by moving `clumpFractions()` into tokenizer.py.

I'll also try to address the issues I'm running into with printing unicode bytestrings without breaking existing functionality of the CLI.

**Comment by wcedmisten** [comment]
@mtlynch I finally fixed the unicode issue I was running into by using `sys.stdout.buffer.write()` instead of `print()`. 

Let me know what you think!

File: bin/parse-ingredients.py [hunk 1]
```diff
@@ -10,12 +10,11 @@
 
 
 def _exec_crf_test(input_text, model_path):
-    with tempfile.NamedTemporaryFile() as input_file:
+    with tempfile.NamedTemporaryFile(mode="w") as input_file:
```
Comment (mtlynch) [inline bin/parse-ingredients.py:13]: Can we do single quotes for consistency?

File: ingredient_phrase_tagger/training/tokenizer.py [hunk 2]
```diff
@@ -1,16 +1,25 @@
+#!/usr/bin/env python3
```
Comment (mtlynch) [inline ingredient_phrase_tagger/training/tokenizer.py:1]: We don't need this, as the shebang line is only for executable files.

This project follows Google Python conventions:

https://github.com/google/styleguide/blob/gh-pages/pyguide.md#37-shebang-line

File: ingredient_phrase_tagger/training/tokenizer.py [hunk 3]
```diff
@@ -1,16 +1,25 @@
...
 
 
 def tokenize(s):
     """
     Tokenize on parenthesis, punctuation, spaces and American units followed by a slash.
-
     We sometimes give American units and metric units for baking recipes. For example:
         * 2 tablespoons/30 mililiters milk or cream
         * 2 1/2 cups/300 grams all-purpose flour
-
```
Comment (mtlynch) [inline ingredient_phrase_tagger/training/tokenizer.py:13]: Can we restore the blank lines on 9 and 13? They don't seem related to the 2-to-3 migration.

File: ingredient_phrase_tagger/training/utils.py [hunk 4]
```diff
@@ -1,25 +1,13 @@
-#!/usr/bin/env python2
+#!/usr/bin/env python3
```
Comment (mtlynch) [inline ingredient_phrase_tagger/training/utils.py:1]: This shouldn't be here in the first place, so we can just delete it.

File: ingredient_phrase_tagger/training/cli.py [hunk 5]
```diff
@@ -14,10 +15,14 @@ def run(self):
         Generates training data in the CRF++ format for the ingredient
         tagging task
         """
-        with open(self.opts.data_path) as data_file:
+        with open(self.opts.data_path, encoding='utf-8') as data_file:
             data_reader = labelled_data.Reader(data_file)
             for row in data_reader:
-                print translator.translate_row(row).encode('utf-8')
+                # write the utf-8 encoded data directly to stdout instead of using print
```
Comment (mtlynch) [inline ingredient_phrase_tagger/training/cli.py:21]: Clear comment, but can we capitalize the first letter and end in a period for consistency?

File: tests/golden/eval_output [hunk 6]
```diff
@@ -2,9 +2,9 @@
 Sentence-Level Stats:
 	correct:  1636
 	total:  2200
-	% correct:  74.3636363636
+	% correct:  74.36363636363636
```
Comment (mtlynch) [inline tests/golden/eval_output:5]: Instead of changing this, could we just change the print call so that it preserves the legacy behavior? I just want to make it super obvious to anyone reading that we didn't change behavior in this PR.

**Review by mtlynch — CHANGES_REQUESTED** [review]
Thanks! Just a few notes, mostly minor style fixes.

**Comment by wcedmisten** [comment]
@mtlynch I believe this addresses all your comments. Let me know if I missed something. Thanks!

**Comment by mtlynch** [comment]
Awesome, thank you!
